### ISSUE TITLE
sae_manifold_fit: atom_basis="duchon" panics with cudarc "Unable to dynamically load cuda" on CPU-only hosts

### ISSUE THREAD

**Issue by SauersML** [SauersML/gam#168]
## Observed
`gamfit.sae_manifold_fit(..., atom_basis="duchon", ...)` aborts the
interpreter with a Rust panic from `cudarc-0.19.7/src/lib.rs:200` even
though the gamfit startup banner already reported

    [GPU] CUDA acceleration disabled
      backend: CPU (faer + Rayon)
      reason: CUDA driver library not found

The panic fires regardless of `assignment_prior` (both `softmax` and
`ibp_map` trigger it), and regardless of `n_atoms` / `atom_dim`. It
only affects `atom_basis="duchon"`; `atom_basis="periodic"` works.

This makes the linear-shard baseline in
`tests/test_sae_manifold_curved_beats_linear.py` (which uses
`atom_basis="duchon", assignment_prior="softmax"`) completely
unrunnable on any CPU-only host (e.g. Apple Silicon, plain Linux
without an NVIDIA driver).

`pyo3_runtime.PanicException` is also notable because it bypasses
ordinary `except Exception` handlers (must be caught with
`BaseException`).

## Expected
- When the CPU backend is selected at startup, code paths must not
  call into `cudarc` at all.
- A missing CUDA driver on CPU-only systems should not cause a Rust
  panic from `sae_manifold_fit`; if a particular feature truly needs
  CUDA it should raise a clean Python exception.

## Reproducer (29 lines)
```python
import math, numpy as np, gamfit
print("gamfit", gamfit.__version__)
rng = np.random.default_rng(0)
n, p = 200, 32
theta = rng.uniform(0, 2*math.pi, n)
harm = np.column_stack([np.cos(theta), np.sin(theta)])
mix = rng.normal(size=(2, p)); mix /= np.linalg.norm(mix, axis=0, keepdims=True)
z = harm @ mix + 0.04 * rng.normal(size=(n, p))
z -= z.mean(0, keepdims=True)

# periodic + ibp_map: OK on CPU
gamfit.sae_manifold_fit(Z=z, n_atoms=1, atom_basis="periodic",
    atom_dim=2, assignment_prior="ibp_map",
    max_iter=20, learning_rate=0.04, random_state=0)
print("periodic ibp_map OK")

# duchon + ibp_map: PanicException (cudarc cannot dlopen libcuda)
gamfit.sae_manifold_fit(Z=z, n_atoms=2, atom_basis="duchon",
    atom_dim=1, assignment_prior="ibp_map",
    max_iter=20, learning_rate=0.04, random_state=0)
```

Panic line:
```
thread '<unnamed>' panicked at cudarc-0.19.7/src/lib.rs:200:5:
Unable to dynamically load the "cuda" shared library ...
pyo3_runtime.PanicException: ...
```

## Environment
- gamfit 0.1.123 (pypi wheel)
- macOS arm64 (Darwin 25.2.0), no NVIDIA driver
- Python 3.13

**Comment by SauersML** [comment]
Fixed on `main` by commit c10e6636 ("gpu/runtime: gate probe on cudarc::is_culib_present to fix macOS panic") and a regression test added in fc16960d.

**Root cause.** `cudarc 0.19`'s entry points lazily initialize `libcuda` through a process-wide `OnceLock<libloading::Library>`. On a host with no CUDA driver (every macOS host, Linux without an NVIDIA runtime), the first call to any cudarc API panics unconditionally from inside `cudarc::lib::panic_no_lib_found` — even when the rest of the program already saw `GpuRuntime::probe()` return `None` and printed the `[GPU] CUDA acceleration disabled` banner. The 0.1.123 PyPI wheel hit this panic on the first cudarc-touching dispatch after probe.

**Fix.** `GpuRuntime::probe()` now gates the driver API on the non-panicking `unsafe { cudarc::driver::sys::is_culib_present() }` probe before calling `CudaContext::device_count()`. When the probe returns `false`, the runtime records the CPU reason, logs the disabled banner, and returns `Err(GpuProbeError::Driver)`. Every public GPU dispatch in the crate routes through `GpuRuntime::global()` (which caches the probe result), so a CPU-only host now never reaches a cudarc sys call:

- `gpu::linalg::route_through_gpu` → all `try_fast_*` / `try_cholesky_*` / `try_solve_*_triangular` entry points
- `gpu::session::upload_x` → `try_fast_xt_diag_x_arc`
- `gpu::blas::*_cuda` → take `runtime: &GpuRuntime` by reference
- `gpu::solver::context_and_stream` → `cholesky_solve_gpu`, `cholesky_lower_gpu`
- `solver::gpu::pirls_gpu::*_gpu` → CPU fallback when runtime is unavailable

**Regression test.** `src/gpu/runtime.rs::tests::cpu_only_host_never_panics_on_gpu_entry_points` asserts that on a host where `is_culib_present()` returns `false`, `GpuRuntime::is_available()` returns `false` and a representative cross-section of GPU entry points (`try_fast_ab`, `try_fast_av`, `try_fast_atv`, `try_cholesky_lower_inplace`, `cholesky_solve_gpu`, `cholesky_lower_gpu`, `pirls_gpu::weighted_crossprod_gpu`) all return cleanly without panicking.

The PyPI 0.1.124 wheel and later carry this fix; 0.1.123 (the version cited in this issue) predates the gate.
